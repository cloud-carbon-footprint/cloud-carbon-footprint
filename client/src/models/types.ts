/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */
import { Account } from './FilterInputModels'

export interface serviceEstimate {
  cloudProvider: string
  accountName: string
  serviceName: string
  wattHours: number
  co2e: number
  cost: number
  region: string
  usesAverageCPUConstant?: boolean
}

export interface ServiceResult {
  data: EstimationResult[]
  loading: boolean
}

export interface EstimationResult {
  timestamp: Date
  serviceEstimates: serviceEstimate[]
}

export interface cloudEstPerDay {
  x: Date
  y: number
  usesAverageCPUConstant?: boolean
  wattHours?: number
  cost?: number
}

export enum ChartDataTypes {
  REGION = 'region',
  SERVICE = 'service',
  ACCOUNT = 'account',
}

export interface FilterResultResponse {
  accounts: Account[]
}
