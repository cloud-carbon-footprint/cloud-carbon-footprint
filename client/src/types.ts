export interface serviceEstimate{
    timestamp: Date,
    serviceName: string,
    wattHours: number,
    co2e: number,
    cost: number
}

export interface ServiceResult{
    data: EstimationResult[],
    loading: boolean,
    error: boolean
}

export interface EstimationResult{
    timestamp:Date,
    serviceEstimates: serviceEstimate[]
}

export interface co2PerDay{
    x: Date,
    y: number
}
