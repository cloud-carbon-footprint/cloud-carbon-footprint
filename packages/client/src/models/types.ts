/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */
import { DropdownOption } from '../dashboard/filters/DropdownFilter'

export interface serviceEstimate {
  cloudProvider: string
  accountName: string
  serviceName: string
  kilowattHours: number
  co2e: number
  cost: number
  region: string
  usesAverageCPUConstant?: boolean
}

export interface ServiceResult<T> {
  data: T[]
  loading: boolean
}

export interface EstimationResult {
  timestamp: Date
  serviceEstimates: serviceEstimate[]
}

export interface cloudEstPerDay {
  x: Date
  y: number
  usesAverageCPUConstant?: boolean
  kilowattHours?: number
  cost?: number
}

export enum ChartDataTypes {
  REGION = 'region',
  SERVICE = 'service',
  ACCOUNT = 'account',
}

export interface FilterResultResponse {
  accounts: DropdownOption[]
  services: DropdownOption[]
}

export enum UnknownTypes {
  UNKNOWN_REGION = 'Unknown Region',
  UNKNOWN_SERVICE = 'Unknown Service',
  UNKNOWN_ACCOUNT = 'Unknown Account',
}

export interface EmissionsRatios {
  region: string
  mtPerKwHour: number
}
