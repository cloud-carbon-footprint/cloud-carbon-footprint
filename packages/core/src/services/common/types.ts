/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

export type QUERY_DATE_TYPES = {
  [key: string]: string
}
