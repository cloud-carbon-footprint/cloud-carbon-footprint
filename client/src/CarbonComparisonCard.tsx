import React, { FunctionComponent, useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import { Card, CardContent, CardActions, Typography, Button } from '@material-ui/core'
import { DriveEta, LocalGasStation, Eco } from '@material-ui/icons'
import { sumCO2 } from './transformData'
import { EstimationResult } from './types'

const useStyles = makeStyles({
  root: {
    width: '100%',
    height: '100%',
  },
  bullet: {
    display: 'inline-block',
    margin: '0 2px',
    transform: 'scale(0.8)',
  },
  title: {
    fontSize: 16,
    color: '#FFFFFF',
  },
  posOne: {
    fontSize: 16,
    color: '#FFFFFF',
    marginBottom: 12,
  },
  posTwo: {
    maxWidth: 250,
  },
  topContainer: {
    backgroundColor: '#2C82BE',
    textAlign: 'center',
  },
  bottomContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    height: '60%',
  },
  buttonContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  metricOne: {
    color: '#FFFFFF',
    fontWeight: 'bold',
  },
  metricTwo: {
    color: '#2C82BE',
    fontWeight: 'bold',
    margin: '20px 0',
  },
  icon: {
    height: 100,
    width: 100,
    color: '#2C82BE',
    padding: 20,
  },
  defaultButton: {
    color: '#000000',
    backgroundColor: '#E0E0E0',
    '&:hover': {
      backgroundColor: '##D5D5D5',
    },
  },
  activeButton: {
    color: '#FFFFFF',
    backgroundColor: '#2C82BE',
    '&:hover': {
    backgroundColor: '#2f75a8',
    },
  },
})

export const CarbonComparisonCard: FunctionComponent<CarbonComparisonCardProps> = ({ data }) => {
  const classes = useStyles()
  const [selection, setSelection] = useState('miles')
  const kgSum: number = sumCO2(data)

  const milesSum: any = (kgSum * 2.48138958).toPrecision(2)
  const gallonsSum: any = (kgSum * .1125239).toPrecision(2)
  const treesSum: any = (kgSum * .0165352).toPrecision(2)


  const comparisons: any = {
    gas: {
          icon: <LocalGasStation className={classes.icon} data-testid='gasIcon' />,
          total: gallonsSum,
          textOne: 'CO2 emissions from',
          textTwo: 'gallons of gasoline consumed',
    },
    miles: {
        icon: <DriveEta className={classes.icon} data-testid='milesIcon' />,
        total: milesSum,
        textOne: 'greenhouse gas emissions from',
        textTwo: 'miles driven by an average passenger vehicle',
    },
    trees: {
        icon: <Eco className={classes.icon}  data-testid='treesIcon' />,
        total: treesSum,
        textOne: 'carbon sequestered by',
        textTwo: 'tree seedlings grown for 10 years',
    },
  }

  const updateSelection = (selection: string) => {
    setSelection(selection)
  }

  let gasButtonClass: any = classes.defaultButton
  let treesButtonClass: any = classes.defaultButton
  let milesButtonClass: any = classes.activeButton

  if (selection === 'gas') {
    gasButtonClass = classes.activeButton
    treesButtonClass = classes.defaultButton
    milesButtonClass = classes.defaultButton
  } else if (selection === 'miles') {
    gasButtonClass = classes.defaultButton
    treesButtonClass = classes.defaultButton
    milesButtonClass = classes.activeButton
  } else if (selection === 'trees') {
    gasButtonClass = classes.defaultButton
    treesButtonClass = classes.activeButton
    milesButtonClass = classes.defaultButton
  }
  
  return (
    <Card className={classes.root}>
        <CardContent className={classes.topContainer}>
            <Typography className={classes.title} gutterBottom>
                Your Cumulative Emissions are
            </Typography>
            <Typography className={classes.metricOne} variant="h4" component="h2">
                {kgSum.toPrecision(2)} kg CO2e
            </Typography>
            <Typography className={classes.posOne}>
                that is equivalent to
            </Typography>
        </CardContent>
        <CardContent className={classes.bottomContainer}>
            <Typography>
                {comparisons[selection].icon}
            </Typography>
            <CardContent>
                <Typography className={classes.posTwo} variant="h5" component="h2">
                    {comparisons[selection].textOne}
                </Typography>
                <Typography className={classes.metricTwo} variant="h3" component="h2">
                    {comparisons[selection].total}
                </Typography>
                <Typography className={classes.posTwo} variant="h5" component="h2">
                    {comparisons[selection].textTwo}
                </Typography>
            </CardContent>
        </CardContent>
        <CardActions className={classes.buttonContainer}>
            <Button variant="contained" className={milesButtonClass} size="medium" onClick={() => updateSelection('miles')}>Miles</Button>
            <Button variant="contained" className={gasButtonClass} size="medium" onClick={() => updateSelection('gas')}>Gas</Button>
            <Button variant="contained" className={treesButtonClass} size="medium" onClick={() => updateSelection('trees')}>Trees</Button>
        </CardActions>
    </Card>
  )
}

type CarbonComparisonCardProps = {
    data: EstimationResult[]
}