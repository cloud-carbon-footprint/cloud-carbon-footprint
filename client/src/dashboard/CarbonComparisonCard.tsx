import React, { FunctionComponent, useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import { Card, CardContent, CardActions, Typography, Button } from '@material-ui/core'
import { DriveEta, LocalGasStation, Eco } from '@material-ui/icons'
import { sumCO2 } from './transformData'
import { EstimationResult } from '../types'

type Selection = 'miles' | 'gas' | 'trees'

type ComparisionItem = {
  icon: React.ReactNode
  total: string
  textOne: string
  textTwo: string
}

type Comparision = {
  [char: string]: ComparisionItem
  gas: ComparisionItem
  miles: ComparisionItem
  trees: ComparisionItem
}

const useStyles = makeStyles(({ palette }) => ({
  root: {
    width: '100%',
    height: '100%',
  },
  title: {
    color: palette.primary.contrastText,
  },
  posOne: {
    color: palette.primary.contrastText,
  },
  posTwo: {
    maxWidth: 250,
  },
  topContainer: {
    backgroundColor: palette.primary.main,
    textAlign: 'center',
  },
  bottomContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    height: '60%',
  },
  buttonContainer: {
    display: 'flex',
    justifyContent: 'center',
  },
  metricOne: {
    color: palette.primary.contrastText,
    fontWeight: 'bold',
  },
  metricTwo: {
    color: palette.primary.light,
    fontWeight: 'bold',
    margin: '1.25rem 0',
  },
  icon: {
    height: 100,
    width: 100,
    color: palette.primary.light,
    padding: '1.25rem',
  },
}))

export const CarbonComparisonCard: FunctionComponent<CarbonComparisonCardProps> = ({ data }) => {
  const classes = useStyles()
  const [selection, setSelection] = useState('miles')
  const kgSum: number = sumCO2(data)

  const milesSum: string = (kgSum * 2.48138958).toPrecision(2)
  const gasSum: string = (kgSum * 0.1125239).toPrecision(2)
  const treesSum: string = (kgSum * 0.0165352).toPrecision(2)

  const comparisons: Comparision = {
    gas: {
      icon: <LocalGasStation className={classes.icon} data-testid="gasIcon" />,
      total: gasSum,
      textOne: 'CO2 emissions from',
      textTwo: 'gallons of gasoline consumed',
    },
    miles: {
      icon: <DriveEta className={classes.icon} data-testid="milesIcon" />,
      total: milesSum,
      textOne: 'greenhouse gas emissions from',
      textTwo: 'miles driven by an average passenger vehicle',
    },
    trees: {
      icon: <Eco className={classes.icon} data-testid="treesIcon" />,
      total: treesSum,
      textOne: 'carbon sequestered by',
      textTwo: 'tree seedlings grown for 10 years',
    },
  }

  const updateSelection = (selection: Selection) => {
    setSelection(selection)
  }

  const updateButtonColor = (buttonSelection: Selection) => {
    return buttonSelection === selection ? 'primary' : 'default'
  }

  return (
    <Card className={classes.root}>
      <CardContent className={classes.topContainer}>
        <Typography className={classes.title} gutterBottom>
          Your Cumulative Emissions are
        </Typography>
        <Typography className={classes.metricOne} variant="h4" component="p">
          {kgSum.toPrecision(2)} kg CO2e
        </Typography>
        <Typography className={classes.posOne}>that is equivalent to</Typography>
      </CardContent>
      <CardContent className={classes.bottomContainer}>
        <Typography>{comparisons[selection].icon}</Typography>
        <CardContent>
          <Typography className={classes.posTwo} variant="h5" component="p">
            {comparisons[selection].textOne}
          </Typography>
          <Typography className={classes.metricTwo} variant="h3" component="p">
            {comparisons[selection].total}
          </Typography>
          <Typography className={classes.posTwo} variant="h5" component="p">
            {comparisons[selection].textTwo}
          </Typography>
        </CardContent>
      </CardContent>
      <CardActions className={classes.buttonContainer}>
        <Button
          variant="contained"
          color={updateButtonColor('miles')}
          size="medium"
          onClick={() => updateSelection('miles')}
        >
          Miles
        </Button>
        <Button
          variant="contained"
          color={updateButtonColor('gas')}
          size="medium"
          onClick={() => updateSelection('gas')}
        >
          Gas
        </Button>
        <Button
          variant="contained"
          color={updateButtonColor('trees')}
          size="medium"
          onClick={() => updateSelection('trees')}
        >
          Trees
        </Button>
      </CardActions>
    </Card>
  )
}

type CarbonComparisonCardProps = {
  data: EstimationResult[]
}
