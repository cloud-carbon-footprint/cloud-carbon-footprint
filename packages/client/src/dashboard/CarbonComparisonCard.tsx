/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React, { FunctionComponent, useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import {
  Card,
  CardContent,
  CardActions,
  Typography,
  Button,
  Link,
} from '@material-ui/core'
import {
  FlightTakeoff,
  PhonelinkRing,
  Eco,
  OpenInNew,
} from '@material-ui/icons'
import { sumCO2 } from './transformData'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import NoDataPage from './NoDataPage'

type Selection = 'flights' | 'phones' | 'trees'

type ComparisonItem = {
  icon: React.ReactNode
  total: number
  textOne: string
  textTwo: string
}

type Comparison = {
  [char: string]: ComparisonItem
  flights: ComparisonItem
  phones: ComparisonItem
  trees: ComparisonItem
}

type SourceItem = {
  href: string
  title: string
}

type Source = {
  [char: string]: SourceItem
  flights: SourceItem
  epa: SourceItem
}

const useStyles = makeStyles(({ palette, spacing, typography }) => {
  return {
    root: {
      width: '100%',
      height: '100%',
      minHeight: '755px',
    },
    title: {
      color: palette.primary.contrastText,
    },
    posOne: {
      color: palette.primary.contrastText,
    },
    posTwo: {
      maxWidth: 250,
    },
    topContainer: {
      backgroundColor: palette.primary.main,
      textAlign: 'center',
    },
    bottomContainer: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      height: '55%',
      paddingTop: '10%',
    },
    buttonContainer: {
      display: 'flex',
      justifyContent: 'center',
    },
    metricOne: {
      color: palette.primary.contrastText,
      fontWeight: typography.fontWeightBold,
    },
    metricTwo: {
      color: palette.primary.light,
      fontWeight: typography.fontWeightBold,
    },
    icon: {
      height: 240,
      width: 200,
      color: palette.primary.light,
    },
    source: {
      padding: spacing(2),
      display: 'flex',
      justifyContent: 'center',
    },
    sourceLink: {
      padding: spacing(0, 1),
      display: 'inline-flex',
      alignItems: 'center',
      color: palette.extLink,
    },
    openIcon: {
      marginLeft: '8px',
    },
    noData: {
      marginTop: '75px',
    },
  }
})

export const toMiles = (co2mt: number): number => co2mt * 2481.3918390475
export const toGas = (co2mt: number): number => co2mt * 112.5247230304
export const toTrees = (co2mt: number): number => co2mt * 16.5337915448

export const CarbonComparisonCard: FunctionComponent<CarbonComparisonCardProps> =
  ({ data }) => {
    const classes = useStyles()
    const [selection, setSelection] = useState('flights')
    const mtSum: number = sumCO2(data)

    const milesSum = toMiles(mtSum)
    const gasSum = toGas(mtSum)
    const treesSum = toTrees(mtSum)

    const formatNumber = (number: number, decimalPlaces = 0) =>
      number.toLocaleString(undefined, { maximumFractionDigits: decimalPlaces })

    const comparisons: Comparison = {
      flights: {
        icon: (
          <FlightTakeoff className={classes.icon} data-testid="flightsIcon" />
        ),
        total: milesSum,
        textOne: 'greenhouse gas emissions from',
        textTwo: 'miles driven on average',
      },
      phones: {
        icon: (
          <PhonelinkRing className={classes.icon} data-testid="phonesIcon" />
        ),
        total: gasSum,
        textOne: 'CO2 emissions from',
        textTwo: 'gallons of gasoline consumed',
      },
      trees: {
        icon: <Eco className={classes.icon} data-testid="treesIcon" />,
        total: treesSum,
        textOne: 'carbon sequestered by',
        textTwo: 'tree seedlings grown for 10 years',
      },
    }

    const sources: Source = {
      flights: {
        href: 'https://calculator.carbonfootprint.com/calculator.aspx?tab=3',
        title: 'Flight Carbon Footprint Calculator',
      },
      epa: {
        href: 'https://www.epa.gov/energy/greenhouse-gas-equivalencies-calculator',
        title: 'EPA Equivalencies Calculator',
      },
    }

    const updateSelection = (selection: Selection) => {
      setSelection(selection)
    }

    const updateButtonColor = (buttonSelection: Selection) => {
      return buttonSelection === selection ? 'primary' : 'default'
    }

    const currentSource = sources[selection] || sources.epa

    return (
      <Card className={classes.root} id="carbonComparisonCard">
        {mtSum ? (
          <div>
            <CardContent className={classes.topContainer}>
              <Typography className={classes.title} gutterBottom>
                Your cumulative emissions are
              </Typography>
              <Typography
                className={classes.metricOne}
                id="metric-one"
                variant="h4"
                component="p"
                data-testid="co2"
              >
                {formatNumber(mtSum, 1)} metric tons CO2e
              </Typography>
              <Typography className={classes.posOne}>
                that is equivalent to
              </Typography>
            </CardContent>
            <CardContent className={classes.bottomContainer}>
              <CardContent>{comparisons[selection].icon}</CardContent>
              <CardContent>
                <Typography
                  className={classes.posTwo}
                  variant="h5"
                  component="p"
                >
                  {comparisons[selection].textOne}
                </Typography>
                <Typography
                  className={classes.metricTwo}
                  variant="h3"
                  component="p"
                  data-testid="comparison"
                >
                  {formatNumber(comparisons[selection].total)}
                </Typography>
                <Typography
                  className={classes.posTwo}
                  variant="h5"
                  component="p"
                >
                  {comparisons[selection].textTwo}
                </Typography>
              </CardContent>
            </CardContent>
            <CardActions className={classes.buttonContainer}>
              <Button
                id="flights"
                variant="contained"
                color={updateButtonColor('flights')}
                size="medium"
                onClick={() => updateSelection('flights')}
              >
                Flights
              </Button>
              <Button
                id="phones"
                variant="contained"
                color={updateButtonColor('phones')}
                size="medium"
                onClick={() => updateSelection('phones')}
              >
                Phones
              </Button>
              <Button
                id="trees"
                variant="contained"
                color={updateButtonColor('trees')}
                size="medium"
                onClick={() => updateSelection('trees')}
              >
                Trees
              </Button>
            </CardActions>
            <Typography className={classes.source} data-testid="epa-source">
              Source:{' '}
              <Link
                href={currentSource.href}
                target="_blank"
                rel="noopener"
                className={classes.sourceLink}
              >
                {currentSource.title}{' '}
                <OpenInNew
                  fontSize={'small'}
                  className={classes.openIcon}
                ></OpenInNew>
              </Link>
            </Typography>
          </div>
        ) : (
          <div>
            <CardContent className={classes.topContainer}>
              <Typography
                className={classes.metricOne}
                variant="h4"
                component="p"
                data-testid="co2"
              >
                Emissions comparison
              </Typography>
            </CardContent>
            <div className={classes.noData}>
              <NoDataPage isTop={false} />
            </div>
          </div>
        )}
      </Card>
    )
  }

type CarbonComparisonCardProps = {
  data: EstimationResult[]
}
