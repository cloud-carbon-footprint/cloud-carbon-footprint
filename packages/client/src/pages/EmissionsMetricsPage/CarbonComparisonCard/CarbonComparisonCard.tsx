/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { FunctionComponent, useState } from 'react'
import { FlightTakeoff, PhonelinkRing, Eco } from '@material-ui/icons'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import { sumEstimates } from '../../../utils/helpers'
import NoDataMessage from '../../../common/NoDataMessage'
import DashboardCard from '../../../layout/DashboardCard'
import { Source, ComparisonItem } from '../../../Types'
import CarbonComparison from './CarbonComparison'
import useStyles from './carbonComparisonStyles'
import { formattedNumberWithCommas } from '../../../utils/helpers/transformData'

type Selection = 'flights' | 'phones' | 'trees'

type Comparisons = {
  [name: string]: ComparisonItem
  flights: ComparisonItem
  phones: ComparisonItem
  trees: ComparisonItem
}

type CarbonComparisonCardProps = {
  data: EstimationResult[]
}

const CarbonComparisonCard: FunctionComponent<CarbonComparisonCardProps> = ({
  data,
}) => {
  const [selection, setSelection] = useState('flights')
  const classes = useStyles()

  const toFlights = (co2mt: number): number => co2mt * 1.2345679 // direct one way flight from NYC to London per metric ton per CO2
  const toPhones = (co2mt: number): number => co2mt * 121643 // phones charged per metric ton of CO2
  const toTrees = (co2mt: number): number => co2mt * 16.5337915448

  const formatNumber = (number: number, decimalPlaces = 0) => {
    if (number >= 1000000000) return `${(number / 1000000000).toFixed(1)}+ B`

    if (number >= 1000000) return `${(number / 1000000).toFixed(1)}+ M`

    return number >= 1
      ? formattedNumberWithCommas(number, decimalPlaces)
      : Number(number.toExponential(decimalPlaces)).toString()
  }

  const totalMetricTons = sumEstimates(data, 'co2e')
  const totalFlights = toFlights(totalMetricTons)
  const totalPhones = toPhones(totalMetricTons)
  const totalTrees = toTrees(totalMetricTons)

  const sources: { [name: string]: Source } = {
    epa: {
      href: 'https://www.epa.gov/energy/greenhouse-gas-equivalencies-calculator',
      title: 'EPA Equivalencies Calculator',
    },
    flightCalculator: {
      href: 'https://calculator.carbonfootprint.com/calculator.aspx?tab=3',
      title: 'Flight Carbon Footprint Calculator',
    },
  }

  const comparisons: Comparisons = {
    flights: {
      icon: (
        <FlightTakeoff className={classes.icon} data-testid="flightsIcon" />
      ),
      total: totalFlights,
      textOne: 'CO2e emissions from',
      textTwo: 'direct one way flights from NYC to London',
      source: sources.flightCalculator,
    },
    phones: {
      icon: <PhonelinkRing className={classes.icon} data-testid="phonesIcon" />,
      total: totalPhones,
      textOne: 'CO2e emissions from',
      textTwo: 'smartphones charged',
      source: sources.epa,
    },
    trees: {
      icon: <Eco className={classes.icon} data-testid="treesIcon" />,
      total: totalTrees,
      textOne: 'carbon sequestered by',
      textTwo: 'tree seedlings grown for 10 years',
      source: sources.epa,
    },
  }

  const updateSelection = (selection: Selection) => {
    setSelection(selection)
  }

  const updateButtonColor = (buttonSelection: Selection) =>
    buttonSelection === selection ? 'primary' : 'default'

  if (totalMetricTons) {
    return (
      <DashboardCard
        isHalf
        noPadding
        testId="carbonComparison"
        id="carbonComparisonCard"
      >
        <CarbonComparison
          formatNumber={formatNumber}
          totalMetricTons={totalMetricTons}
          comparisons={comparisons}
          selection={selection}
          updateSelection={updateSelection}
          updateButtonColor={updateButtonColor}
        />
      </DashboardCard>
    )
  }

  return <NoDataMessage isHalf boldTitle="Emissions Comparison" />
}

export default CarbonComparisonCard
