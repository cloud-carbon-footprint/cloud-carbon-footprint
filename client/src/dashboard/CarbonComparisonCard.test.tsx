/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React from 'react'
import { create } from 'react-test-renderer'
import moment from 'moment'
import { render, fireEvent } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import { CarbonComparisonCard, toGas, toMiles, toTrees } from './CarbonComparisonCard'
import { EstimationResult } from '../models/types'

const data = [
  {
    timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
    serviceEstimates: [
      {
        serviceName: 'ebs',
        wattHours: 0,
        co2e: 5,
        cost: 0,
        region: 'us-west-2',
      },
      {
        serviceName: 's3',
        wattHours: 0,
        co2e: 0,
        cost: 0,
        region: 'us-west-2',
      },
      {
        serviceName: 'ec2',
        wattHours: 0,
        co2e: 2,
        cost: 0,
        region: 'us-west-2',
      },
    ],
  },
]

describe('CarbonComparisonCard', () => {
  it('renders with correct configuration', () => {
    const root = create(<CarbonComparisonCard data={data} />)

    expect(root.toJSON()).toMatchSnapshot()
  })

  it('selects carbon comparison correctly', async () => {
    const { getByText, getByTestId } = render(<CarbonComparisonCard data={data} />)

    const gasButton = getByText('Gas')
    const milesIcon = getByTestId('milesIcon')
    expect(gasButton).toBeInstanceOf(HTMLElement)
    expect(milesIcon).toBeInstanceOf(SVGSVGElement)

    act(() => {
      fireEvent.click(gasButton)
    })

    const gasIcon = getByTestId('gasIcon')
    expect(gasIcon).toBeInstanceOf(SVGSVGElement)

    const treesButton = getByText('Trees')
    expect(treesButton).toBeInstanceOf(HTMLElement)

    act(() => {
      fireEvent.click(treesButton)
    })

    const treesIcon = getByTestId('treesIcon')
    expect(treesIcon).toBeInstanceOf(SVGSVGElement)

    const milesButton = getByText('Miles')
    expect(milesButton).toBeInstanceOf(HTMLElement)

    act(() => {
      fireEvent.click(milesButton)
    })

    expect(milesIcon).toBeInstanceOf(SVGSVGElement)
  })

  it('should show EPA source', async () => {
    const { getByTestId } = render(<CarbonComparisonCard data={data} />)
    const source = getByTestId('epa-source')
    expect(source).toHaveTextContent('Source: EPA Equivalencies Calculator')
  })

  it('should open EPA page in other tab when clicking EPA link', async () => {
    const { getByText } = render(<CarbonComparisonCard data={data} />)
    const epaLink = getByText('EPA Equivalencies Calculator').closest('a')
    expect(epaLink).toHaveAttribute('href', 'https://www.epa.gov/energy/greenhouse-gas-equivalencies-calculator')
    expect(epaLink).toHaveAttribute('target', '_blank')
    //for security reasons https://web.dev/external-anchors-use-rel-noopener/
    expect(epaLink).toHaveAttribute('rel', 'noopener')
  })

  describe('number formatting', () => {
    let data: EstimationResult[]
    let co2kg: number
    beforeEach(() => {
      data = [
        {
          timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
          serviceEstimates: [
            {
              serviceName: 'ebs',
              wattHours: 0,
              co2e: 10000,
              cost: 0,
              region: 'us-west-2',
            },
          ],
        },
      ]
      co2kg = data[0].serviceEstimates[0].co2e
    })

    it('should format co2e to default locale integer', async () => {
      const co2e = 999999.55555555555
      data[0].serviceEstimates[0].co2e = co2e
      const { getByTestId } = render(<CarbonComparisonCard data={data} />)
      const co2 = getByTestId('co2')
      expect(co2).toHaveTextContent(co2e.toLocaleString(undefined, { maximumFractionDigits: 0 }))
    })

    it('should format miles', async () => {
      const { getByTestId } = render(<CarbonComparisonCard data={data} />)
      const co2 = getByTestId('comparison')
      const expected = toMiles(co2kg).toLocaleString(undefined, { maximumFractionDigits: 0 })
      expect(co2).toHaveTextContent(expected)
    })

    it('should format gas', async () => {
      const { getByText, getByTestId } = render(<CarbonComparisonCard data={data} />)
      act(() => {
        fireEvent.click(getByText('Gas'))
      })
      const co2 = getByTestId('comparison')
      const expected = toGas(co2kg).toLocaleString(undefined, { maximumFractionDigits: 0 })
      expect(co2).toHaveTextContent(expected)
    })

    it('should format trees', async () => {
      const { getByText, getByTestId } = render(<CarbonComparisonCard data={data} />)
      act(() => {
        fireEvent.click(getByText('Trees'))
      })
      const co2 = getByTestId('comparison')
      const expected = toTrees(co2kg).toLocaleString(undefined, { maximumFractionDigits: 0 })
      expect(co2).toHaveTextContent(expected)
    })
  })
})
