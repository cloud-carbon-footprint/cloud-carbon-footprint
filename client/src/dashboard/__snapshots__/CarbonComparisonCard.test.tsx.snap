// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarbonComparisonCard renders with correct configuration 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardContent-root makeStyles-topContainer-5"
  >
    <p
      className="MuiTypography-root makeStyles-title-2 MuiTypography-body1 MuiTypography-gutterBottom"
    >
      Your cumulative emissions are
    </p>
    <p
      className="MuiTypography-root makeStyles-metricOne-8 MuiTypography-h4"
      data-testid="co2"
    >
      7
       kg CO2e
    </p>
    <p
      className="MuiTypography-root makeStyles-posOne-3 MuiTypography-body1"
    >
      that is equivalent to
    </p>
  </div>
  <div
    className="MuiCardContent-root makeStyles-bottomContainer-6"
  >
    <div
      className="MuiCardContent-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icon-10"
        data-testid="milesIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18.92 5.01C18.72 4.42 18.16 4 17.5 4h-11c-.66 0-1.21.42-1.42 1.01L3 11v8c0 .55.45 1 1 1h1c.55 0 1-.45 1-1v-1h12v1c0 .55.45 1 1 1h1c.55 0 1-.45 1-1v-8l-2.08-5.99zM6.5 15c-.83 0-1.5-.67-1.5-1.5S5.67 12 6.5 12s1.5.67 1.5 1.5S7.33 15 6.5 15zm11 0c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zM5 10l1.5-4.5h11L19 10H5z"
        />
      </svg>
    </div>
    <div
      className="MuiCardContent-root"
    >
      <p
        className="MuiTypography-root makeStyles-posTwo-4 MuiTypography-h5"
      >
        greenhouse gas emissions from
      </p>
      <p
        className="MuiTypography-root makeStyles-metricTwo-9 MuiTypography-h3"
        data-testid="comparison"
      >
        17
      </p>
      <p
        className="MuiTypography-root makeStyles-posTwo-4 MuiTypography-h5"
      >
        miles driven by an average passenger vehicle
      </p>
    </div>
  </div>
  <div
    className="MuiCardActions-root makeStyles-buttonContainer-7 MuiCardActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Miles
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Gas
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Trees
      </span>
    </button>
  </div>
  <p
    className="MuiTypography-root makeStyles-source-11 MuiTypography-body1"
    data-testid="epa-source"
  >
    Source:
     
    <a
      className="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-sourceLink-12 MuiTypography-colorPrimary"
      href="https://www.epa.gov/energy/greenhouse-gas-equivalencies-calculator"
      onBlur={[Function]}
      onFocus={[Function]}
      rel="noopener"
      target="_blank"
    >
      EPA Equivalencies Calculator 
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-openIcon-13 MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
        />
      </svg>
    </a>
  </p>
</div>
`;
